File: sportsbook-app/betting-service/src/main/java/com/sportsbook/bettingservice/service/BetService.java
package com.sportsbook.bettingservice.service;

import com.sportsbook.bettingservice.domain.Bet;
import com.sportsbook.bettingservice.domain.BetStatus;
import com.sportsbook.bettingservice.kafka.BetEventMapper;
import com.sportsbook.bettingservice.kafka.BetEventProducer;
import com.sportsbook.bettingservice.kafka.BetPlacedEvent;
import com.sportsbook.bettingservice.repo.BetRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
@RequiredArgsConstructor
public class BetService {

    private final BetRepository betRepository;
    private final BetEventProducer producer;
    private final BetEventMapper eventMapper;

    /**
     * Keep your current signature. Interpret stake as currency units (e.g. 25.00),
     * and derive stakeCents for persistence.
     */
    @Transactional
    public Bet placeBet(String playerId, String marketId, double stake, double odds) {
        Bet bet = Bet.builder()
                .id(UUID.randomUUID().toString())
                .playerId(playerId)
                .marketId(marketId)
                .stake(stake)
                .stakeCents(stake * 100d)
                .odds(odds)
                .status(BetStatus.PENDING)
                .build();

        betRepository.save(bet);

        // Map domain -> local event via MapStruct, then publish
        BetPlacedEvent event = eventMapper.toLocalEvent(bet);
        producer.publishBetPlaced(event);

        return bet;
    }
}

File: sportsbook-app/betting-service/src/main/kotlin/com/sportsbook/bettingservice/service/BetService.kt
package com.sportsbook.bettingservice.service

import com.sportsbook.bettingservice.domain.Bet
import com.sportsbook.bettingservice.domain.BetStatus
import com.sportsbook.bettingservice.kafka.BetEventMapper
import com.sportsbook.bettingservice.kafka.BetEventProducer
import com.sportsbook.bettingservice.repo.BetRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Service
class BetService(
    private val betRepository: BetRepository,
    private val producer: BetEventProducer,
    private val eventMapper: BetEventMapper,
) {

    @Transactional
    fun placeBet(playerId: String, marketId: String, stake: Double, odds: Double): Bet {
        val bet = Bet().apply {
            id = UUID.randomUUID().toString()
            this.playerId = playerId
            this.marketId = marketId
            this.stake = stake
            this.stakeCents = stake * 100.0
            this.odds = odds
            this.status = BetStatus.PENDING
        }

        betRepository.save(bet)

        val event = eventMapper.toLocalEvent(bet)
        producer.publishBetPlaced(event)

        return bet
    }
}
